gittyup: Fix build with GCC 14

phijor opened the pull request:
The test suite of gittyup depends on a vendored version of a ZIP library (https://github.com/kuba--/zip), and the build breaks when compiled with GCC 14.  This is fixed in an unreleased commit (https://github.com/Murmele/Gittyup/commit/d36eba172a01d541945d59427b4f643aaed55da0) upstream.

~~Fix the build by disabling the test suite.  In a future release, the test suite should be re-enabled.~~

Patch the affected file and leave a note to remove the patch in a future release.

This fixes #371778, and (imo) supersedes #371808. @FliegendeWurst can you try and see if this the build failure you mention in https://github.com/NixOS/nixpkgs/pull/371808#issuecomment-2575957466? 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

phijor committed:
gittyup: Prefer using `cmakeBool`

FliegendeWurst commented:
I would prefer setting NIX_CFLAGS_COMPILE to `-Wno-error=calloc-transposed-args`

FliegendeWurst commented:
Build looks mostly good https://paste.fliegendewurst.eu/d4-v4d.log#L2789

I enabled strictDeps by default. Would be good to move cmark to nativeBuildInputs so it doesn't break in that configuration.

phijor committed:
gittyup: Fix generation of HTML docs from Markdown files

`cmark` is needed to generate documentation at build time.

phijor committed:
gittyup: Fix build with GCC 14

GCC 14 turns transposed arguments to `calloc()` into an error.  The
affected file has been fixed upstream; patch it here to make things
work until the next release of `gittyup`.

phijor commented:
> I would prefer setting NIX_CFLAGS_COMPILE to `-Wno-error=calloc-transposed-args`

I have considered doing this, but instead have decided to simply patch the affected file. That's cleaner than silencing the error. I guess this is also easier to maintain; the patch will likely no longer apply when the next upstream release makes it into `nixpkgs`, and at that point it can be removed again.

FliegendeWurst commented on pkgs/applications/version-management/gittyup/default.nix:
```diff
@@ -29,29 +29,35 @@ stdenv.mkDerivation rec {
-    "-DUSE_SYSTEM_OPENSSL=ON"
-    "-DENABLE_UPDATE_OVER_GUI=OFF"
-  ];
+  patches = [ ./0001-Fix-incorrect-order-of-argument-to-calloc-345.patch ];
```
```suggestion
  patches = [
    # Fix GCC 14 build error (remove for next update)
    # https://github.com/Murmele/Gittyup/pull/759
    ./0001-Fix-incorrect-order-of-argument-to-calloc-345.patch
  ];
```

FliegendeWurst commented on pkgs/applications/version-management/gittyup/default.nix:
```diff
@@ -29,29 +29,35 @@ stdenv.mkDerivation rec {
 
   buildInputs =
     [
-      cmark
```
Actually I just checked again. It seems cmark is both required as build tool and regular library. So this needs to stay.

phijor committed:
gittyup: add fliegendewurst and phijor as maintainers

phijor commented:
> You could also add yourself, if you would like to help out with issues and testing updates :) Even as a casual user your input would be much appreciated.

Done! In general I try to maintain _less_ CMake-based projects, but life doesn't always go that way :D I'm happy to help out.
